Added element-by-id, elements-by-class-name, and elements-by-tag-name finders

JaimeIvanCervantes opened the pull request:
Here is a list of changes in this PR:
* Add `element_by_id` finder and cache
* Add `elements_by_class_name` finder and cache
* Add `elements_by_tag_name` finder and cache
* Renamed to `style_stack_by_id` finder and cache, as part of the finder refactoring
* Replaced `map` with `unordered_map` in style-stack cache (this improves search by 2x on my computer)
* Removed `element_info` and replaced with `style_stack` (the `element` is already returned by one of the finders above)
* Refactored `CacheCreator` to adhere to the style standards of the rest of the codebase 

Also, I am not convinced that `style_stack_by_id` caching should be part of the finder, maybe it should be a separate class? I like the current approach though, because all the caches are populated by a single visitor pass.

Also, here are a few TODOs:
* Update `svgren` to support changes to finder
* Maybe, populate the caches on the Finder initializer list

JaimeIvanCervantes commented:
also, I will be very busy in the next days @igagis, so feel free to add anything that might be missing in the meantime

igagis commented:
Hi Jamie,

thanks for efforts. The code looks good and it sound great that unordered_map improves performance 2 times.

But I think there was small misunderstanding regarding how to organize the code.
I meant that for each search type (i.e. by id, by tag, by class) there would be a separate finder class. I.e. `class finder_by_id`, `class finder_by_tag`, `class finder_by_class`, three separate C++ classes, instead of putting all the functionality into one existing `finder` class. And thus, the existing `finder` class remains unchanged (maybe except that `unordered_map` thingy). Could you split your implementation to those three separate classes? Each class should be in separate `.hpp`/`.cpp` files.

Also, since there will be only one 'find' and one 'size' method per such class, there is no need to name those that specific, those method names can be just `find(...)` and `size()`.

Also, separate tests should be added for each such class.

igagis commented:
I'll update `svgren` later, after new `svgdom` with the changes is released. The `svgren` changes would be very simple, just some renamings.

igagis commented on src/svgdom/util/finder.cpp:
```diff
@@ -7,107 +7,189 @@
 public:
-	std::map<std::string, finder::element_info> cache;
+	std::unordered_map<std::string, const element*> element_by_id_cache;
+	std::unordered_map<std::string, std::list<const element*>> elements_by_class_name_cache;
```
use `std::vector` instead of `std::list`
See https://www.youtube.com/watch?v=YQs6IC-vgmo

igagis commented on src/svgdom/util/finder.cpp:
```diff
@@ -7,107 +7,189 @@
+				it->second.push_back(&e);
+			} else {
+				std::list<const element*> elements = {&e};
+				elements_by_class_name_cache.insert(std::make_pair(class_name, elements));
```
use `std::move(elements)` here to avoid unnecessary copying

igagis commented on src/svgdom/util/finder.cpp:
```diff
@@ -7,107 +7,189 @@
+	}
+
+	void add_to_elements_by_tag_name_cache(const svgdom::element& e, const svgdom::styleable& s){
+		if (!s.get_tag().empty()) {
```
actually the `get_tag()` method should probably be also a part of `svgdom::element`. And, thus, we would not need reference to `styleable` here. Because every element has a tag, not only styleable elements. So, we need to introduce the virtual `element::get_tag()`. I have just did it, see latest commit on master branch. Just merge the latest master to your forked repo.

igagis commented on src/svgdom/util/finder.hpp:
```diff
@@ -12,29 +12,49 @@ class finder{
 private:
-	std::map<std::string, element_info> cache;
+	std::unordered_map<std::string, const element*> element_by_id_cache;
+	std::unordered_map<std::string, std::list< const element*>> elements_by_class_name_cache;
```
same here, use `std::vector`

igagis commented:
also, I see that you have changed the `back.svg` file. Please, if you need more test data, then add new files, but do not modify existing ones. So, in this case, please copy the original `back.svg` and apply your changes to the copy.

JaimeIvanCervantes commented:
I addressed most of the comments, separated the finders, and renamed the old finder to `style_stack_cache`. I still have to add a new svg and tests for the finders, and revert `back.svg` and `back.svg.cmp`. I'll do this over the weekend.

igagis commented on src/svgdom/util/element_by_id_finder.cpp:
```diff
@@ -0,0 +1,114 @@
+		}
+	}
+	
+	void visit_container(const svgdom::element& e, const svgdom::container& c, const svgdom::styleable& s){
```
in sthis case we are not interested in `styleable` part of the element, thus, the `styleable` argument can be removed from this function. And then, at a first glance it looks to me that the visitor becomes same as standard visitor. I mean that in this case you don't have to override every `visit()` method of the class, maybe only overriding of `default_visit()` could be enough, thus shortening the code significantly. Please have a look if that is possible in this case.

igagis commented on src/svgdom/util/elements_by_class_name_finder.cpp:
```diff
@@ -0,0 +1,115 @@
+public:
+	std::unordered_map<std::string, std::vector<const element*>> cache;
+
+	void add_to_cache(const svgdom::element& e, const svgdom::styleable& s){
```
this method could be made `private:`, just a minor thing

igagis commented on src/svgdom/util/elements_by_tag_name_finder.cpp:
```diff
@@ -0,0 +1,115 @@
+	}
+	
+	void visit_container(const svgdom::element& e, const svgdom::container& c, const svgdom::styleable& s){
+		this->add_to_cache(e, s);
```
same thing here, we don't need `styleable` argument, the `get_tag()` method is now present in base `element` class.

igagis commented on src/svgdom/util/style_stack_cache.cpp:
```diff
@@ -7,101 +7,101 @@
-	style_stack styleStack;
-	
-	void addToCache(const svgdom::element& e){
+	style_stack style_stack;
```
add namespace to type name, i.e. `svgdom::style_stack style_stack` as the name of the variable and class name are same in this case, compiler probably will give an error.
Or, better solution is to avoid this kind of name clashing and use different name for the variable, e.g. `style_stack current_style_stack`, in this case the variable name is more descriptive and more precisely describes the meaning of the variable.

igagis commented:
> I addressed most of the comments

small advice, for those addressed comments you can press the "Resolve conversation" button

igagis commented on src/svgdom/util/element_by_id_finder.cpp:
```diff
@@ -0,0 +1,114 @@
+#include "element_by_id_finder.hpp"
```
by the way, why did you choose that long name? Maybe we could make the name a bit shorter, because it is self evident that we are searching for `element`s, thus it is not needed to have the 'element' word in the name. Also, it makes the namings more inconsistent, as in case of class we need to indicate that more than one element can be found, which is also self evident and is also already indicated by the `find()` method's signature. This is why I proposed the `finder_by_id`, `find_by_class`, `find_by_tag` names. Do those names sound unnatural to native English speaker? Maybe it could be also `by_id_finder` but I think it is better to use Polish notation, when we first type the most general entity, i.e. `finder` and then qualify it more by adding `by_id`, but I'm not sure, at least to my ear it sounds more clear in this case.

JaimeIvanCervantes replied:
I used those name to be consistent with the Javascript methods, and denote that `id` returns a single element, and `class-name` and `tag-name` return one or more elements. I though about using `find_element()` and `find_elements()` as an alternative.

igagis replied:
Well, I don't think we need to  blindly follow Javascript namings here, C++ is a different language and has its specifics.
Javascript is dynamically typed, so there is no other way to indicate the return type of the method, other than saying that in method's name. In C++ we have strict types, so from the method signature we see what type it returns, if it is supposed to return multiple elements, then we see the return type as some container, e.g. `std::vector`, there is no chance to mix it with something else, as compiler will point you to the mistake in compile time (thanks to strict typing). So, adding notion to method's name is redundant information. Also, to avoid unnecessary typing we need to keep names as short as possible, while preserving evidence of what method/class does. So, as I mentioned, it is very unlikely that in context of `svgdom` we'd be searching for something else than `element`s, so I see no need to include the `element` word in the namings. Same about `tag`/`class` vs `tag_name`/`class_name`, the `name` part adds no value to the descriptiveness of the method, when one says just `tag` it is quite clear that we mean it's name, and not, for example length of the tag.
So, I still would like the names would be `finder_by_id`, `finder_by_class`, `finder_by_tag`. Could you do the renamings please?

JaimeIvanCervantes commented:
I pushed the test for the finders and addressed all comments....

igagis commented on src/svgdom/util/elements_by_class_name_finder.hpp:
```diff
@@ -0,0 +1,30 @@
+
+	elements_by_class_name_finder(const svgdom::element& root);
+
+	const std::vector<const element*> find(const std::string& cls)const;
```
I think this method should return reference to vector to avoid copying, i.e. `const std::vector<const element*>&`

igagis commented on src/svgdom/util/elements_by_tag_name_finder.hpp:
```diff
@@ -0,0 +1,30 @@
+
+	elements_by_tag_name_finder(const svgdom::element& root);
+
+	const std::vector<const element*> find(const std::string& tag_name)const;
```
same here, should return reference to a vector

igagis commented on src/svgdom/util/style_stack_cache.cpp:
```diff
@@ -0,0 +1,116 @@
+#include "style_stack_cache.hpp"
```
By the way, do you use this `style_stack_cache` class somewhere in your programs? I'm thinking about moving it from `svgdom` to `svgren` in the future, possibly. In case it is too specific thing which is not needed for common use. Just asking, no need to do anything about it so far.

JaimeIvanCervantes replied:
I think it's important to keep `style_stack_cache` because it's not only used for rendering; if one would like to find bounding boxes for all elements to manipulate them, for example, the style-stacks would have to be used to determine the true position and bounds.

igagis replied:
Ok, I'll keep it in `svgdom`.

igagis commented:
The PR is almost OK now, still a few minor thing to correct. Thanks for the efforts!

JaimeIvanCervantes commented:
I addressed the second round of comments

JaimeIvanCervantes commented:
I had to revert the element vectors by reference `const std::vector<const svgdom::element*>`, because the CI was failing with this error:
```
error: returning reference to temporary
```
When I return the list initializer `{}` in the case that `class` or `tag` is not found.

igagis commented:
> I had to revert the element vectors by reference

right, I had a closer look at this and it turns out that using `std::vector` is not the right thing in this case at all. I have changed the code to use `span`.

I had also improved the finder tests.

igagis reviewed (approved):
The PR looks OK now

igagis commented:
Thanks a lot for the efforts. I will release a new version of `svgdom` shortly and than I'll start updating the `svgren` accordingly.

JaimeIvanCervantes commented:
Glad to help! And I do a lot of PR reviews, so this was easy 😉, and also helpful to understand your style. This way I can help in the future.

One last question @igagis, why use you `span` instead of `std::vector`?

igagis commented:
@JaimeIvanCervantes The `span` in this case is, as I mentioned before, to avoid unnecessary creation of the `vector`'s copy.
Our finder object works in a way that it first creates a cache in it's constructor and after that the cache remains unchanged during the object's life time. So, we only need to able able to read the cache data, and making a copy of the cache data's portion every time we need to read that portion of data is far not optimal. The `span` is actually what is called a `view` of the data's portion. So, we only return the pointer to the memory and length of the memory area relevant, without actually copying it.
Returning `std::vector` would result in memory allocation (relatively slow operation) and then copying the memory to the allocated memory. So, why doing it if we don't actually need it and can avoid it? :)

Since C++'17 we have `std::string_view` which provides the `view` facility for strings. But for arbitrary arrays of data they are introducing `std::span` only in C++'20 (see https://en.cppreference.com/w/cpp/container/span). So, for the time being I maintain my own implementation `utki::span` which will become deprecated when C++'20 standard is well supported by compilers and I move my libs to C++'20. But so far, C++'20 standard is not well supported by compilers, many implementations are still experimental.
